posthog/test/base.py

```diff
@@ -557,10 +557,30 @@ def execute_wrapper(query, *args, **kwargs):
             yield queries
 
 
+class RaiseOnJoinThread(threading.Thread):
+    """
+    By default, the thread will swallow exceptions and not raise them in the
+    main thread. This class will raise them in the main thread such that we do
+    not fail silently e.g. on failures to apply ClickHouse schemas.
+    """
+
+    def run(self):
+        self.exc = None
+        try:
+            super().run()
+        except BaseException as e:
+            self.exc = e
```

Comment:
I appreciate all work related to making errors less-silent* :handshake: 

That being said, I'm not sure how are things supposed to work if more than 1 Thread fails. Seems like we'll be inevitably swallowing exceptions as only the first one will be processed (re-raised).

I -think- I would have personally overridden `threading.excepthook` to either:

1. Handle whatever I can handle in the Thread context.
2. Explicitly fail immediately (by re-raising) if I run into an unrecoverable error that makes all other work redundant.
3. Store the error for later processing if we can continue working on other threads (some care would be required when storing objects to avoid reference cycles). Maybe a thread that failed can be retried, or the other threads rolled-back.

* (nit) By default, everything that is not `SystemExit` is being logged to stderr, so I wouldn't say it's completely silent.

Reply:
True not completely silent. If all the tests pass despite mutation failures, it is silent as pytest is capturing output. If a test failed then you would see the stderr output.

The rough behaviour I was going for was something similar to if these were all run sequentially in the same thread where we'd likely have a fail fast approach. The main thing is that the error causes the tests to fail.

For 1. I'm not sure there's anything to handle. It failed and therefore we should fail the test run.
For 2. I wasn't aware that re-raising in excepthook would cause the main thread to halt but that sounds like a good option.
For 3. I think if we start needing to handle retries in tests then our tests are too fragile, but I could be being a bit dogmatic there.

What do you think the best option is? I'll add fixes to https://github.com/PostHog/posthog/pull/13820 🙇